File: src/main/java/priv/bajdcc/LALR1/semantic/tracker/Tracker.java
package priv.bajdcc.LALR1.semantic.tracker;

import priv.bajdcc.LALR1.syntax.automata.npa.NPAStatus;
import priv.bajdcc.util.lexer.regex.IRegexStringIterator;

import java.util.Stack;

/**
 * 跟踪器（链表）
 *
 * @author bajdcc
 */
public class Tracker {
	/**
	 * 指令记录集
	 */
	public InstructionRecord rcdInst = null;

	/**
	 * 错误记录集
	 */
	public ErrorRecord rcdError = null;

	/**
	 * 当前PDA状态
	 */
	public NPAStatus npaStatus = null;
	/**
	 * PDA状态堆栈
	 */
	public Stack<NPAStatus> stkStatus = new Stack<>();

	/**
	 * 单词遍历接口
	 */
	public IRegexStringIterator iter = null;

	/**
	 * 是否产生了错误
	 */
	public boolean bRaiseError = false;

	/**
	 * 当前步骤是否产生了错误
	 */
	public boolean bInStepError = false;

	/**
	 * 是否已经中止
	 */
	public boolean bFinished = false;

	/**
	 * 前向指针
	 */
	public Tracker prev = null;

	/**
	 * 后向指针
	 */
	public Tracker next = null;
}

File: src/main/kotlin/com/bajdcc/LALR1/semantic/tracker/Tracker.kt
package com.bajdcc.LALR1.semantic.tracker

import com.bajdcc.LALR1.syntax.automata.npa.NPAStatus
import com.bajdcc.util.lexer.regex.IRegexStringIterator
import java.util.*

/**
 * 跟踪器（链表）
 *
 * @author bajdcc
 */
class Tracker {
    /**
     * 指令记录集
     */
    var insts: InstructionRecord? = null

    /**
     * 错误记录集
     */
    var errors: ErrorRecord? = null

    /**
     * 当前PDA状态
     */
    var npaStatus: NPAStatus? = null
    /**
     * PDA状态堆栈
     */
    var stkStatus = Stack<NPAStatus>()

    /**
     * 单词遍历接口
     */
    var iter: IRegexStringIterator? = null

    /**
     * 是否产生了错误
     */
    var raiseError = false

    /**
     * 当前步骤是否产生了错误
     */
    var inStepError = false

    /**
     * 是否已经中止
     */
    var finished = false

    /**
     * 前向指针
     */
    var prev: Tracker? = null

    /**
     * 后向指针
     */
    var next: Tracker? = null
}
